### PR TITLE
#162685666 Users can comment on articles

### PR DESCRIPTION
**Pull request by Opio-Emmanuel-Omona**
#### What does this PR do?
This PR enables only authenticated users to comment on an article(s).

#### Description of Task to be completed?
Authenticated users are able to comment on articles that have been published. The users are able to perform all the CRUD functionality on the comments of the articles.
Available endpoints include
```
GET api/<slug>/comments/         Gets all comments for a particular article
```
```
POST api/<slug>/comments/         Creates a comment for a particular article
```
```
GET api/<slug>/comments/<id>/         Get a specific comment of an article
```
```
PUT api/<slug>/comments/<id>/         Update a specific comment of an article
```
```
DELETE api/<slug>/comments/<id>/         Delete a specific comment of an article
```
```
POST api/<slug>/comments/<id>/thread/         Comment on a comment
```
```
GET api/<slug>/comments/<id>/thread/         Get entire thread of a comment
```

#### How should this be manually tested?
Run the following commands in your terminal.

- `git clone https://github.com/andela/Ah-backend-guardians.git`
- `cd Ah-backend-guardians`
- `virtualenv venv`
- `source venv/bin/activate`
- `git checkout ft-comment-on-articles-162685666`
- `python3 manage.py runserver`

Alternatively, you can test the app using the Heroku link below
https://ah-backend-guardians-sta-pr-25.herokuapp.com/

Next open POSTMAN and begin to test out the end points listed above.
The data required in a POST and PUT is in the form
```
{
	"body":  ""
}
```

Note: You must be logged in to access the endpoints and atleast one article should be present in your database.

#### What are the relevant pivotal tracker stories?
#162685666

#### Screenshots (if appropriate)
This shows a GET all comments for an article
image

### CODE REVIEW

**Review by fahadmak — CHANGES_REQUESTED** [review]
it is not necessary for you to have input ` parent: id` key-value pair in your post. yet you have the parent id in the urls you pass for creating a thread comment

screen shot 2019-01-22 at 11 39 33 am

**Comment by Opio-Emmanuel-Omona** [comment]
@fahadmak I have worked on the "parent" key issue.
screen shot 2019-01-22 at 12 18 17 pm

**Review by moseskamira — APPROVED** [review]
LGTM. I have reviewed it both locally and on Heroku too.

**Comment by fahadmak** [comment]
LGTM

**Review by fahadmak — APPROVED** [review]

**Comment by archbishop12** [comment]
@Opio-Emmanuel-Omona Ensure one cannot delete/update a comment he/she did not create.

**Comment by archbishop12** [comment]
@Opio-Emmanuel-Omona Can threaded comments be updated or deleted???

**Comment by archbishop12** [comment]
@Opio-Emmanuel-Omona Run pep8 on the files you added to ensure there are no violations.

File: authors/apps/comments/views.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        slug = self.kwargs['slug']
+        try:
+            article = Article.objects.get(slug=slug)
+            try:
+                comment = Comment.objects.get(id=id)
+            except Comment.DoesNotExist:
+                message = 'Comment does not exist'
+                return Response(message, status.HTTP_404_NOT_FOUND)
+        except Article.DoesNotExist:
+            message = 'Article does not exist'
```
Comment (archbishop12) [inline authors/apps/comments/views.py]: @Opio-Emmanuel-Omona Same applies here.

File: authors/apps/comments/views.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        }, status=status.HTTP_200_OK)
+
+    def destroy(self, request, id, *args, **kwargs):
+        slug = self.kwargs['slug']
+        try:
+            article = Article.objects.get(slug=slug)
+            try:
+                comment = Comment.objects.get(id=id)
+            except Comment.DoesNotExist:
+                message = 'Comment does not exist'
```
Comment (archbishop12) [inline authors/apps/comments/views.py]: @Opio-Emmanuel-Omona Return a dictionary here to maintain consistency through out the entire application.
```suggestion
                message = {'message': 'Comment does not exist'}
```

File: authors/apps/comments/views.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        slug = self.kwargs['slug']
+        try:
+            article = Article.objects.get(slug=slug)
+            try:
+                parent = Comment.objects.get(pk=id)
+            except Comment.DoesNotExist:
+                message = 'Comment does not exist'
+                return Response(message, status.HTTP_404_NOT_FOUND)
+        except Article.DoesNotExist:
+            message = 'Article does not exist'
```
Comment (archbishop12) [inline authors/apps/comments/views.py]: @Opio-Emmanuel-Omona Same applies here.

File: authors/apps/comments/views.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    queryset = Comment.objects.all()
+
+    def post(self, request, id, *args, **kwargs):
+        slug = self.kwargs['slug']
+        try:
+            article = Article.objects.get(slug=slug)
+            try:
+                parent = Comment.objects.get(pk=id)
+            except Comment.DoesNotExist:
+                message = 'Comment does not exist'
```
Comment (archbishop12) [inline authors/apps/comments/views.py]: @Opio-Emmanuel-Omona Same applies here.

File: authors/apps/comments/views.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+                return Response(message, status.HTTP_404_NOT_FOUND)
+        except Article.DoesNotExist:
+            message = 'Article does not exist'
```
Comment (archbishop12) [inline authors/apps/comments/views.py]: @Opio-Emmanuel-Omona Same applies here.

File: authors/apps/comments/views.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+                        status=status.HTTP_201_CREATED)
+
+    def get(self, request, id, *args, **kwargs):
+        slug = self.kwargs['slug']
+        try:
+            article = Article.objects.get(slug=slug)
+            try:
+                parent = Comment.objects.get(pk=id)
+            except Comment.DoesNotExist:
+                message = 'Comment does not exist'
```
Comment (archbishop12) [inline authors/apps/comments/views.py]: @Opio-Emmanuel-Omona Same applies here.

File: authors/apps/comments/views.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+        slug = self.kwargs['slug']
+        try:
+            article = Article.objects.get(slug=slug)
+            try:
+                comment = Comment.objects.get(id=id)
+            except Comment.DoesNotExist:
+                message = 'Comment does not exist'
+                return Response(message, status.HTTP_404_NOT_FOUND)
+        except Article.DoesNotExist:
+            message = 'Article does not exist'
```
Comment (archbishop12) [inline authors/apps/comments/views.py]: @Opio-Emmanuel-Omona Same applies here.

File: authors/apps/comments/views.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        return Response({"comment": serializer.data})
+
+    def update(self, request, id, *args, **kwargs):
+        slug = self.kwargs['slug']
+        try:
+            article = Article.objects.get(slug=slug)
+            try:
+                comment = Comment.objects.get(id=id)
+            except Comment.DoesNotExist:
+                message = 'Comment does not exist'
```
Comment (archbishop12) [inline authors/apps/comments/views.py]: @Opio-Emmanuel-Omona Same applies here.

File: authors/apps/comments/views.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        serializer.save()
+        return Response({"message": "Comment Successfully added"},
+                        status=status.HTTP_201_CREATED)
+
+    def get(self, *args, **kwargs):
+        slug = self.kwargs['slug']
+        try:
+            article = Article.objects.get(slug=slug)
+        except Article.DoesNotExist:
+            message = 'Article does not exist'
```
Comment (archbishop12) [inline authors/apps/comments/views.py]: @Opio-Emmanuel-Omona Some of these messages seem to appear multiple times. Could you avoid repeating yourself.

**Review by fahadmak — CHANGES_REQUESTED** [review]
Add docstring comments in your tests and comments endpoint

**Comment by Opio-Emmanuel-Omona** [comment]
@archbishop12 

> @Opio-Emmanuel-Omona Can threaded comments be updated or deleted???
Yes, they can, They have IDs and are deleted/updated just like any other comment. 
```
DELETE api/<slug>/comments/<id>/
```

> @Opio-Emmanuel-Omona Ensure one cannot delete/update a comment he/she did not create.
Resolved: Only the author of the comment can delete/update the comment


> @Opio-Emmanuel-Omona Run pep8 on the files you added to ensure there are no violations.
Resolved: All files follow the pep8 guidelines

**Comment by Opio-Emmanuel-Omona** [comment]
@fahadmak 
> Add docstring comments in your tests and comments endpoint

Resolved

**Review by fahadmak — APPROVED** [review]
LGTM

**Comment by archbishop12** [comment]
@Opio-Emmanuel-Omona When one tries to delete a threaded comment using this route `api/<slug>/comments/<id>/` and there exists a non-threaded comment with the same `id` both are deleted. Look into this.

**Review by archbishop12 — APPROVED** [review]
LGTM @Opio-Emmanuel-Omona I am merging this.

**Pull request merged**